### PR TITLE
Add a benchmarking module to wrench.

### PR DESCRIPTION
**Pull request by glennw**
This adds two new commands to wrench:

perf - This runs all the benchmarks in the benchmarks manifest, and records
the timing of various profile counters into a JSON file. This is
very similar to how the reftest code works.

compare_perf - This allows comparing two files output from the perf command above.

There are two new workflows enabled by this change:

(1) Working locally:
 - Run `perf` to get a baseline benchmark.
 - Develop new feature or optimization.
 - Run `perf` to get new results, and use `compare_perf` to compare to baseline.

(2) Automated performance measurement:
 - This is not available yet, but the idea is to upload the results
   of the perf command to a web server, which graphs the results
   over time.

New benchmarks can be added to the benchmarks/ directory, in a similar
way to adding reftests.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/webrender/969)

### CODE REVIEW

**Comment by glennw** [comment]
@kvark Of particular note in this PR - making sure you agree with the general approach I've taken for recording statistics. The current strategy is:

* Keep rendering until we have at least 10 recorded frame profiles (for each of CPU and GPU).
* Sort the profiles array by timings.
* Remove the two fastest frames and the two slowest frames.
* Calculate the mean time from the remaining samples.

I'm completely open to changing that method if think something else would be more suitable.

**Comment by kvark** [comment]
@glennw 
> Calculate the mean time from the remaining samples.

This should be the arithmetic average at this step (instead of mean). It's not robust to extreme samples, but that's why we filter out the 2 highest and 2 lowest in the earlier step.

**Comment by glennw** [comment]
@kvark Ummm... isn't the arithmetic average the same as the mean?

**Comment by kvark** [comment]
Oops, you are correct. I was thought mean=middle. Lost in translation.

File: wrench/src/perf.rs [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+
+    fn render_yaml(&mut self, filename: &Path) -> TestProfile {
+        let mut reader = YamlFrameReader::new(filename);
+        reader.do_frame(self.wrench);
+
+        // wait for the frame
+        self.rx.recv().unwrap();
+
+        // Loop until we get a reasonable number of CPU and GPU
+        // frame profiles. Then take the median.
```
Comment (kvark) [inline wrench/src/perf.rs]: huh, maybe I just read your mind? :) Comment needs to  be updated

File: wrench/src/perf.rs [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+              gpu_frame_profiles.len() < 10 {
+            self.wrench.render();
+            self.window.swap_buffers();
+            let (cpu_profiles, gpu_profiles) = self.wrench.get_frame_profiles();
+            cpu_frame_profiles.extend(cpu_profiles);
+            gpu_frame_profiles.extend(gpu_profiles);
+        }
+
+        cpu_frame_profiles.sort_by(|a, b| {
+            assert!(a.draw_calls == b.draw_calls);
```
Comment (kvark) [inline wrench/src/perf.rs]: nit: there is `assert_eq`

File: wrench/src/perf.rs [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+        while cpu_frame_profiles.len() < 10 ||
+              gpu_frame_profiles.len() < 10 {
+            self.wrench.render();
+            self.window.swap_buffers();
+            let (cpu_profiles, gpu_profiles) = self.wrench.get_frame_profiles();
+            cpu_frame_profiles.extend(cpu_profiles);
+            gpu_frame_profiles.extend(gpu_profiles);
+        }
+
+        cpu_frame_profiles.sort_by(|a, b| {
```
Comment (kvark) [inline wrench/src/perf.rs]: nit: I think `sort_by_key` is simpler

File: wrench/src/perf.rs [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+        gpu_frame_profiles.sort_by(|a, b| {
+            a.paint_time_ns.cmp(&b.paint_time_ns)
+        });
+
+        // Remove the two slowest and fastest frames.
+        let cpu_samples = &cpu_frame_profiles[2..cpu_frame_profiles.len() - 2];
+        let gpu_samples = &gpu_frame_profiles[2..gpu_frame_profiles.len() - 2];
+
+        // Use the mean value from the remaining frames.
+        let composite_time = cpu_samples.iter()
```
Comment (kvark) [inline wrench/src/perf.rs]: nit: could do `map` followed by `sum`

**Comment by glennw** [comment]
Thanks for the review - comments addressed.

**Comment by kvark** [comment]
Thanks!
@bors-servo r+

**Pull request merged**